### PR TITLE
GUpdate status time for messages and reports automatically.

### PR DESCRIPTION
**Pull request by cbrwizard**
# Description

Refactors `relative-time` and related filters into directives. Now the
time automatically updates for them.

medic/medic-webapp#2753

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

File: .jshintrc [hunk 1]
```diff
@@ -64,6 +64,7 @@
     "devel"         : true,     // Development/debugging (alert, confirm, etc)
     "dojo"          : false,    // Dojo Toolkit
     "jquery"        : true,     // jQuery
+    "mocha"        : true,     // Mocha.js
```
Comment (cbrwizard) [inline .jshintrc]: That's needed so that `context` is ok for jshint in tests.

File: static/js/services/live-list.js [hunk 2]
```diff
@@ -108,12 +109,15 @@ angular.module('inboxServices').factory('LiveListConfig',
...
           var element = renderTemplate(scope);
+          element = $(element);
           if (removedDomElement &&
               removedDomElement.find('input[type="checkbox"]').is(':checked')) {
             // updating an item that was selected in select mode
-            element = $(element);
             element.find('input[type="checkbox"]').prop('checked', true);
           }
+          // So that relative-date directive is rendered inside a 'content_row_list_item' template.
+          $compile(element.contents())(scope);
```
Comment (cbrwizard) [inline static/js/services/live-list.js]: I've spent quite a lot of time on this. Had to open a [fiddle](http://jsfiddle.net/cbrwizard/4e2r2o07/) to understand that it's not Angular's fault hehe. Without this all directives inside `content_row_list_item` don't get rendered.

File: Changes.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ _TBD_
 - [#3597](https://github.com/medic/medic-webapp/issues/3597): Update fonts to Noto.
 - [#3561](https://github.com/medic/medic-webapp/issues/3561): Display required fields when creating a "restricted to a place" user.
 - [#2522](https://github.com/medic/medic-webapp/issues/2522): Percentage target values are confusing.
+- [#2753](https://github.com/medic/medic-webapp/issues/2753): Update status time for messages and reports automatically.
```
Comment (cbrwizard) [inline Changes.md]: Is this correct?

**Comment by cbrwizard** [comment]
If there are any comments, I'd love to respond & fix where needed!

The code removes some filters. Can this change be considered breaking? Or it's kinda internal api related?

**Comment by cbrwizard** [comment]
I'm not sure if the failures are related. It fails on `protractor:default`. Here they are:

```
Failures:
1) registration transition submits new sms messages shows content
  Message:
[31m    Expected '  to +64271234567' to be ' to +64271234567'.[0m
  Stack:
    Error: Failed expectation
        at checkAutoResponse (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:233:90)
        at Object.it (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:255:7)
        at /home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:112:25
        at new ManagedPromise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:1067:7)
        at ControlFlow.promise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2396:12)
        at schedulerExecute (/home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:95:18)
        at TaskQueue.execute_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2970:14)
        at TaskQueue.executeNext_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2953:27)
        at asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2860:25)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7
  Message:
[31m    Expected '  to +64271234567' to be ' to +64271234567'.[0m
  Stack:
    Error: Failed expectation
        at checkScheduledTask (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:241:93)
        at Object.it (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:256:7)
        at /home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:112:25
        at new ManagedPromise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:1067:7)
        at ControlFlow.promise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2396:12)
        at schedulerExecute (/home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:95:18)
        at TaskQueue.execute_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2970:14)
        at TaskQueue.executeNext_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2953:27)
        at asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2860:25)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7
  Message:
[31m    Expected '  to +64271234567' to be ' to +64271234567'.[0m
  Stack:
    Error: Failed expectation
        at checkScheduledTask (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:241:93)
        at Object.it (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/registration-by-sms.js:257:7)
        at /home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:112:25
        at new ManagedPromise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:1067:7)
        at ControlFlow.promise (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2396:12)
        at schedulerExecute (/home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:95:18)
        at TaskQueue.execute_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2970:14)
        at TaskQueue.executeNext_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2953:27)
        at asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2860:25)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7

2) sms-gateway api - gateway submits new WT sms messages - shows content
  Message:
[31m    Failed: Element taking too long to appear in the DOM. Giving up!
    Wait timed out after 12008ms[0m
  Stack:
    TimeoutError: Element taking too long to appear in the DOM. Giving up!
    Wait timed out after 12008ms
        at WebDriverError (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/error.js:27:5)
        at TimeoutError (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/error.js:238:5)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2107:17
        at ManagedPromise.invokeCallback_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:1366:14)
        at TaskQueue.execute_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2970:14)
        at TaskQueue.executeNext_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2953:27)
        at asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2813:27)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7
        at process._tickCallback (internal/process/next_tick.js:109:7)
    From: Task: Element taking too long to appear in the DOM. Giving up!
        at scheduleWait (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2094:20)
        at ControlFlow.wait (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2408:12)
        at thenableWebDriverProxy.wait (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/webdriver.js:943:29)
        at run (/home/travis/build/medic/medic-webapp/node_modules/protractor/lib/browser.ts:66:27)
        at ProtractorBrowser.to.(anonymous function) [as wait] (/home/travis/build/medic/medic-webapp/node_modules/protractor/lib/browser.ts:74:12)
        at browser.wait.then.catch (/home/travis/build/medic/medic-webapp/tests/protractor/helper.js:50:24)
        at ManagedPromise.invokeCallback_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:1366:14)
        at TaskQueue.execute_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2970:14)
        at TaskQueue.executeNext_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2953:27)
        at asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2813:27)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7
        at process._tickCallback (internal/process/next_tick.js:109:7)
    From: Task: Run it("- shows content") in control flow
        at Object.<anonymous> (/home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:94:19)
        at /home/travis/build/medic/medic-webapp/node_modules/jasminewd2/index.js:64:48
        at ControlFlow.emit (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/events.js:62:21)
        at ControlFlow.shutdown_ (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2565:10)
        at shutdownTask_.MicroTask (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2490:53)
        at MicroTask.asyncRun (/home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:2619:9)
        at /home/travis/build/medic/medic-webapp/node_modules/selenium-webdriver/lib/promise.js:676:7
        at process._tickCallback (internal/process/next_tick.js:109:7)
    From asynchronous test: 
    Error
        at Suite.describe (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/sms-gateway.js:155:5)
        at Suite.describe (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/sms-gateway.js:142:3)
        at Object.<anonymous> (/home/travis/build/medic/medic-webapp/tests/protractor/e2e/sms-gateway.js:125:1)
        at Module._compile (module.js:570:32)
        at Object.Module._extensions..js (module.js:579:10)
        at Module.load (/home/travis/build/medic/medic-webapp/node_modules/coffee-script/lib/coffee-script/register.js:45:36)
        at tryModuleLoad (module.js:446:12)
        at Function.Module._load (module.js:438:3)
        at Module.require (module.js:497:17)
        at require (internal/module.js:20:19)
        at /home/travis/build/medic/medic-webapp/node_modules/jasmine/lib/jasmine.js:93:5
        at Array.forEach (native)
        at Jasmine.loadSpecs (/home/travis/build/medic/medic-webapp/node_modules/jasmine/lib/jasmine.js:92:18)
        at Jasmine.execute (/home/travis/build/medic/medic-webapp/node_modules/jasmine/lib/jasmine.js:197:8)
        at /home/travis/build/medic/medic-webapp/node_modules/protractor/built/frameworks/jasmine.js:132:15
        at Function.promise (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:682:9)
        at /home/travis/build/medic/medic-webapp/node_modules/protractor/built/frameworks/jasmine.js:104:14
        at _fulfilled (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:834:54)
        at self.promiseDispatch.done (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:863:30)
        at Promise.promise.promiseDispatch (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:796:13)
        at /home/travis/build/medic/medic-webapp/node_modules/q/q.js:604:44
        at runSingle (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:137:13)
        at flush (/home/travis/build/medic/medic-webapp/node_modules/q/q.js:125:13)
        at _combinedTickCallback (internal/process/next_tick.js:73:7)
        at process._tickCallback (internal/process/next_tick.js:104:9)

```

I can't run `grunt e2e` locally to check it. It fails differently:
```
Running "exec:deploytest" (exec) task
added 298 packages in 76.803s
added 139 packages in 18.283s
>> Exited with code: 7.
>> Error executing child process: Error: Process exited with code 7.
```

File: Changes.md (hunk 3 above)
Comment (garethbowen) [reply Changes.md]: It won't make it in to 2.14.0 as we've already branched to `2.14.x`. We usually generate release notes just before release so you can just leave this out for now.

File: static/js/directives/time-ago-auto-update.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      $interval.cancel(timeoutId);
+    });
+
+    // Render it immediately.
+    renderTime(element, momentDate, options);
+
+    // start the UI update process; save the timeoutId for canceling
+    timeoutId = $interval(function() {
+      renderTime(element, momentDate, options);
+    }, 1000);
```
Comment (garethbowen) [inline static/js/directives/time-ago-auto-update.js]: I'm concerned about performance here as we could have 100s of these firing every second and we run on very low end devices. Raising it to once per minute would be a better compromise.

Alternatively the interval could increase as the time gets further away, for example, if the date is two weeks away checking once a day would be fine. If it's variable it may need $timeout instead of $interval so it can be updated each time.

Is there a way of batching this together so we don't get 100s of digest cycles per second? Or will angular look after that for us under the hood?

File: static/js/directives/time-ago-auto-update.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
+var moment = require('moment');
+
+/**
+ * Directive for an auto-updating relative time ago date.
+ * Renders only text and auto-updates it. For better display should be
+ * used in other directives.
+ */
+angular.module('inboxDirectives').directive('timeAgoAutoUpdate', ['$interval', 'FormatDate', function($interval, FormatDate) {
+  'use strict';
+  'ngInject';
```
Comment (garethbowen) [inline static/js/directives/time-ago-auto-update.js]: If you use 'ngInject' you don't need to specify the array of dependencies, so you can just use:

```js
angular.module('inboxDirectives').directive('timeAgoAutoUpdate', function($interval, FormatDate) {
```

If you felt like it, also update the other directives to use 'ngInject'.

File: static/js/services/live-list.js (hunk 2 above)
Comment (garethbowen) [reply static/js/services/live-list.js]: We had some serious performance issues using angular for the lists so we essentially wrote our own renderer with jquery. We're planning to migrate off angularjs to angular/react/something soon and my hope is that will be sufficiently fast so we can use it for lists too.

I think your change will reintroduce the performance issue because it's going back to using angular bindings (if I understand it correctly). LiveList already has custom code to update the dates so it's the one place where your change isn't required. Can you revert this and leave LiveList working as it did before?

File: templates/directives/autoreply.html [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<span ng-if="hasState">
+  <task-state state="task.state"></task-state>
+  &nbsp;
```
Comment (garethbowen) [inline templates/directives/autoreply.html]: This looks like a hack - why is this necessary? There are some below too...

File: templates/partials/content_row_list_item.html [hunk 7]
```diff
@@ -22,7 +26,9 @@
           <i class="fa {{valid ? 'green-dot' : 'red-dot'}} {{verified ? 'fa-check-circle' : 'fa-circle'}}"></i>
         </div>
       </div>
-      <p ng-bind-html="dob | age"></p>
+      <p>
+        <relative-date date="dob" type="age"></relative-date>
+      </p>
       <div class="simprints-stars tier-{{simprintsTier}} {{simprintsTier ? '' : 'ng-hide'}}"></div>
       <div class="detail {{lineage && lineage.length ? '' : 'ng-hide'}}" ng-bind-html="lineage | lineage"></div>
       <div class="warning">{{warning}}</div>
```
Comment (garethbowen) [inline templates/partials/content_row_list_item.html]: As mentioned above, we need to stick with our current non-angular approach for now which can only handle expressions and `ng-bind-html` and nothing else.

**Comment by garethbowen** [comment]
> The code removes some filters. Can this change be considered breaking? Or it's kinda internal api related?

Yes that could be breaking. Our product is highly configurable including the ability to define how information is presented on screen using filters. We have some [configuration documentation](https://github.com/medic/medic-docs/blob/master/configuration/contact-summary.md#configuration) available if you're interested... The bottom line is I think we need to keep the filters around too.

> I can't run grunt e2e locally to check it. It fails differently:

I don't know what that error means off the top of my head. It should run locally. Does the dev environment work? Make sure you have the webdriver running.

Most of those errors look to be just whitespace so you should be able to fix them without running the tests locally. The other one I'm not sure about but you might be able to see what's going on if you follow through the test code.

File: static/js/directives/time-ago-auto-update.js (hunk 5 above)
Comment (cbrwizard) [reply static/js/directives/time-ago-auto-update.js]: Neat

File: static/js/directives/time-ago-auto-update.js (hunk 4 above)
Comment (cbrwizard) [reply static/js/directives/time-ago-auto-update.js]: Yeah, that's a valid concern. I've moved that to a single entity, which handles this and future global timers.

File: templates/directives/autoreply.html (hunk 6 above)
Comment (cbrwizard) [reply templates/directives/autoreply.html]: It's actually copied from a filter https://github.com/medic/medic-webapp/pull/3998/files#diff-59129c7e90017db0b53628a428717201L82. I've changed that now so it looks more human-friendly

File: templates/partials/content_row_list_item.html (hunk 7 above)
Comment (cbrwizard) [reply templates/partials/content_row_list_item.html]: Bummer! I hope migration to React starts soon hehe

**Comment by cbrwizard** [comment]
@garethbowen please check out the alternative solution with a single interval.
Also I've resolved the first failure, but still having issues with e2e tests.

File: static/js/services/recurring-processes-manager.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          return {
+            interval: 1000 * 60,
+            callback: DomMutators.updateRelativeTime
+          };
+        default:
+          throw new Error('Unknown recurring process name');
+      }
+    }
+
+    function startProcessByName(name, options) {
```
Comment (garethbowen) [inline static/js/services/recurring-processes-manager.js]: I don't see the benefit of having named processes. Why not just have `startUpdateRelativeTime` and `stopUpdateRelativeTime` functions?

File: static/js/services/recurring-processes-manager.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    'use strict';
+    'ngInject';
+
+    var recurringProcesses = {};
+
+    function getProcessOptionsByName(name) {
+      switch (name) {
+        case UPDATE_RELATIVE_TIME:
+          return {
+            interval: 1000 * 60,
```
Comment (garethbowen) [inline static/js/services/recurring-processes-manager.js]: I think 10 minutes may be better - we have potentially quite a few dates on screen and very low end hardware so I'm willing to trade off date accuracy for performance.

File: static/js/controllers/messages-content.js [hunk 10]
```diff
@@ -235,5 +242,8 @@ angular.module('inboxControllers').controller('MessagesContentCtrl',
       }
     });
 
+    RecurringProcessesManager.startProcessByName(
+      UPDATE_RELATIVE_TIME
+    );
```
Comment (garethbowen) [inline static/js/controllers/messages-content.js]: I think it might be better to just register/destroy this in InboxCtrl because most pages will have a relative date somewhere on them and that means we won't miss a registration or destroy step on one page.

File: templates/directives/full-date.html [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<div ng-if-start="date" class="relative-date-content">
```
Comment (garethbowen) [inline templates/directives/full-date.html]: If I'm reading this correctly this behaves the same as just `ng-if="date"` - is that right?

File: tests/karma/unit/filters/relative-date.js [hunk 12]
```diff
@@ -42,7 +42,7 @@ describe('relativeDate filter', function() {
     chai.expect(element.text()).to.equal('somerelativetime');
   });
 
-  it('should render a time when the date is today', () => {
+  it('should render a time when the date is today', function() {
```
Comment (garethbowen) [inline tests/karma/unit/filters/relative-date.js]: This is a step backwards. Ideally this PR would not contain changes in any files which are purely whitespace or formatting to try and keep the git history relevant. I suspect some of this is done by your automatic formatter.

File: tests/karma/unit/filters/relative-day.js [hunk 13]
```diff
@@ -3,13 +3,13 @@ describe('relativeDay filter', function() {
...
-      date = sinon.stub(),
-      relative = sinon.stub();
+    scope,
+    date = sinon.stub(),
+    relative = sinon.stub();
 
   beforeEach(function() {
     module('inboxApp');
-    module(function ($provide) {
+    module(function($provide) {
```
Comment (garethbowen) [inline tests/karma/unit/filters/relative-day.js]: Revert these changes too.

File: static/js/filters/date.js [hunk 14]
```diff
@@ -160,16 +80,6 @@ var _ = require('underscore'),
...
-  module.filter('fullDate', function(FormatDate) {
-    return function (date) {
-      if (!date) {
-        return '';
-      }
-      return '<div class="relative-date-content">' + FormatDate.relative(date) + '</div>' +
-             '<div class="full-date">' + FormatDate.datetime(date) + '</div>';
-    };
-  });
-
```
Comment (garethbowen) [inline static/js/filters/date.js]: I think we need to keep this filter around for backwards compatibility.

**Comment by garethbowen** [comment]
Thanks @cbrwizard ! It's definitely getting close.

Those e2e tests don't give enough information unfortunately. You should be able to run them locally and turn off headless mode - if you put sleeps in the right place and inspect the DOM you should be able to see what's going wrong. If not, just clean up the last few things and let me know and I can do a dive to get them running and then we can get this merged!

**Comment by garethbowen** [comment]
Sorry it took so long to get back into this, and now it's been replaced by https://github.com/medic/medic-webapp/pull/4189

Thanks for your hard work @cbrwizard !
